### ISSUE TITLE
type of Octokit is incorrect

### ISSUE THREAD

**Issue by SleeplessByte** [octokit/octokit.js#2115]
TL;DR don't union `void` if you want TypeScript to work as expected.

My original findings in #2112 follow below:

---------------

@gr2m I checked your PR on @neenjaw repro and yes it solves his issue, but I did find the actually issue with the type. 

The type of `Octokit` is, regardless of what Tim does in his repro:

```typescript
Octokit 
  & void 
  & { paginate: PaginateInterface } 
  & Api
```

This results in your PR working because `octokit.graphql` reaches into `Octokit` as can be seen here:

![Autocomplete of type information when typing octokit with the Octokit type showing the grapql property, among others](https://user-images.githubusercontent.com/1964376/121212367-928b4500-c87d-11eb-8fd5-7599bc4d2386.png)

So far, so good, right?

The problem is that `void` that's there. This makes it so that you _can't_ narrow this value when doing a truthy (or falsy) check, which is what Tim is running into. Let me demonstrate:

![Showing that a thruthy check narrows the type to never](https://user-images.githubusercontent.com/1964376/121212874-07f71580-c87e-11eb-86f3-9fb5f474498c.png)

```typescript
const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN! })

if (octokit) {
  octokit.graphql // nope
}
```

This is understandable behaviour. Something that is `void` _can never be truthy_. We can further proof that this is what happens by inspecting the else case:

![Showing that the falsy case has the expected thruthy type](https://user-images.githubusercontent.com/1964376/121213063-3543c380-c87e-11eb-8f85-a8e85c6b4dbe.png)

```typescript
if (octokit) {
} else {
  octokit.graphql // yep!
}
```

The source `octokit.d.ts` has this faulty type:

```typescript
export declare const Octokit: typeof OctokitCore & import("@octokit/core/dist-types/types").Constructor<void & {
    paginate: import("@octokit/plugin-paginate-rest").PaginateInterface;
} & import("@octokit/plugin-rest-endpoint-methods/dist-types/types").Api>;
```

@neenjaw the reason this worked with earlier versions is that `void` didn't trump other types in the union, and now it does. As far as I can tell this has been the change that broke the type, but the change is correct. It shouldn't be `void & { shape }` but just be `{ shape }`.

_Originally posted by @SleeplessByte in https://github.com/octokit/octokit.js/issues/2112#issuecomment-856868654_

**Comment by gr2m** [comment]
I've no idea where the `void` is coming from there. It's quite a deep rabbit hole. I also don't see the use case ochecking if `octokit` is truthy right after you instantiate it? Why would you do that? 

I agree the `void` should be removed, but unless it has a bigger impact I won't give this a high priority

**Comment by SleeplessByte** [comment]
You would do that if you want to pass `octokit` in as a variable, like:


```typescript
function myThing(ocktokit: Octokit): void
function myThing(ocktokit: undefined): void
function myThing(ocktokit: null): void
function myThing(): void

function myThing(ocktokit?: Octokit | null | undefined): void {

}
```

...among many more use cases. Right now we can't pass octokit around unless we assume that it's always there or cast the type manually. 

This also means you can't safely type a conditional initializer or memoize it.

**Comment by gr2m** [comment]
The problem was a plugin returning `void`. I fixed that now, but will also look into preventing that from happening again in the future. Sorry for the trouble. A new release for `octokit` is on its way

**Comment by gr2m** [comment]
The problem should not happen again once https://github.com/octokit/core.js/pull/356 is merged an released. However, I couldn't figure out a way to test for `octokit` being an intersection with `void`. I added a test to `test/typescript-validate.ts`, see https://github.com/octokit/core.js/pull/356/files, but that test passed before already.

Do you have an idea how to add a typescript-only check whether `octokit` has the intersection with void? I'd like preventing that from happening again

**Comment by SleeplessByte** [comment]
You could use the narrowing trick I used:

```typescript
const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN! })

if (octokit) {
  octokit // should be the correct type, so `expectType` here. Should NOT be never
}
```

But I don't know if that will work with the test suite you have set up. When I must _force_ type checkes like this, I usually use what the DefinitelyTyped repo uses.

**Comment by gr2m** [comment]
Thanks! This did the trick

```ts
expectType<void>(octokitWithVoidAndNonVoidPlugins);
```

That worked before octokit/core.js#356 but is now throwing a TypeScript error 👍🏼
